### ISSUE TITLE
Command runs but where is the clang-tidy output and the failures

### ISSUE THREAD

**Issue by jenisys** [lmapii/run-clang-tidy#16]
I use `run-clang-tidy` on a project where `clang-tidy` emits some warnings.
When I use `run-clang-tidy`, I get:

```bash
$ run-clang-tidy .clang-tidy.json 
 [ 1/6 ] Found tidy file /.../clang-tidy-mistakes/.clang-tidy
 [ 2/6 ] Using build root /.../clang-tidy-mistakes/build.multi
 [ 3/6 ] Found 22 files for the provided path patterns
 [ 4/6 ] Found clang-tidy version 16.0.3 using command /usr/local/Cellar/llvm/16.0.3/bin/clang-tidy
 Encountered existing tidy file /.../clang-tidy-mistakes/.clang-tidy
 [ 5/6 ] Existing tidy file matches /.../clang-tidy-mistakes/.clang-tidy, skipping placement
 [ 6/6 ] Executing clang-tidy ...

          Ok build.multi/CMakeFiles/3.26.3/CompilerIdCXX/CMakeCXXCompilerId.cpp
          Ok false-negative/const-assignment.cpp
          Ok false-negative/function-cast.cpp
          Ok false-negative/inifite-goto.cpp
          Ok false-negative/destructor-exception.cpp
          Ok false-negative/stack-corruption.cpp
          Ok false-negative/global-null.cpp
          Ok false-negative/long-loop.cpp
          Ok false-negative/data-execution.cpp
          Ok false-negative/buffer-overrun.cpp
          Ok false-negative/infinite-recursion.cpp
          Ok false-negative/float-loop.cpp
          Ok false-negative/dependency-loop.cpp
          Ok false-negative/virtual-call-constructor.cpp
          Ok false-negative/noexcept-crash.cpp
          Ok false-negative/null-escaping.cpp
          Ok false-negative/virtual-div-by-zero.cpp
          Ok false-negative/memcpy-buffer-overrun.cpp
          Ok false-negative/unhandled-exception.cpp
          Ok false-negative/duplicated-delete.cpp
          Ok false-negative/thread-recursion.cpp
          Ok false-negative/thread-invalid-stack.cpp
     Running [==========================] 22/22                                                                                                                                                                 Finished in 11 seconds
```

by using `.clang_tidy.json` file

```txt
{
    "paths": [
      "**/*.cpp"
    ],
    "buildRoot": "build.multi",
    "tidyFile": ".clang-tidy",
    "tidyRoot": ".",
    "command": "/usr/local/opt/llvm/bin/clang-tidy"
}
```

and a simple `CMakeLists.txt` file to build a library from the source and provide a `build.multi/compile_commands.json` file.

**EXPECTED:**
* Some indication that `clang-tidy` fails (or emits warnings), like when using `clang-tidy` directly (instead of: `Ok`)
* Some output-file(s) with the clang-tidy output/warnings.

**SEE ALSO:**

* Using project example: https://github.com/polystat/clang-tidy-mistakes

**Comment by lmapii** [comment]
Hi, I'll definitely have a look. So far the behaviour of `run-clang-tidy` is very simple: If `clang-tidy` returns a status code of zero, then all the captured output is discarded. Only for any other status code it will raise an error and print whatever was captured on `stout` and `sterr`.

Thus, plain warnings are currently ignored by `run-clang-tidy`. I might have some time on the weekend or latest in CW22 to have a proper go at it. The example project is definitely interesting.

An easy (short-term) solution might be to use `WarningsAsErrors` in your `.clang-tidy` file, alternatively I guess I could also provide `--warning-as-errors` command line or configuration option and pass that to `clang-tidy`.

**Comment by lmapii** [comment]
I've set up the `clang-tidy-mistakes` project on my side. Would you mind sharing your `.clang-tidy` file?

**Comment by jenisys** [comment]
I am using just a simple `.clang-tidy` config-file.
Even if you enable all checks, you will normally only get warnings 
(even so some bad things are in the sources, like a division-by-zero errors, etc.).

EXAMPLE: .clang-tidy

```
---
Checks: "*"
WarningsAsErrors: ""
HeaderFilterRegex: ""
FormatStyle:     none
...
```

NOTES:
* With `WarningsAsErrors: '*'` (as: all warnings should be errors), I get the output as you described.
* With `WarningsAsErrors: "bugprone-*"`, some files have errors.
* Not all problems seem to get caught by `clang-tidy` only. 
  If I use [codechecker](https://github.com/Ericsson/codechecker) that runs `clang-tidy` and `cppcheck`,
  some additional errors are detected, like: other division-by-zero errors, infinite-recursions, etc. (in: `long-loop.cpp`, `dependency-loop.cpp`)

**Comment by lmapii** [comment]
So what I can offer is to adapt `run-clang-tidy` to also display warnings. The warnings could then be suppressed when using `--quiet` or when passing `--ignore-warnings` (this would then have the same behaviour as it has now).

I'd not, however, modify the return code of `run-clang-tidy` in case of warnings. Meaning: If there are warnings and you want a return code that is not zero, then you'll need to use the `clang-tidy` option `WarningsAsErrors`, e.g., when using `run-clang-tidy` in your CI.

Would that help? In terms of what `clang-tidy` can detect or not I can't really help you. This is just a wrapper to selectively execute `clang-tidy` on a subset of your code base. I am not planning to add additional checkers such as `cppcheck` thus far.

**Comment by jenisys** [comment]
* Seeing the the warnings or knowing that warnings exist is my primary concern
* An improvement would be if the status in the file processing progress bar would have 3 outcomes: `Ok, Error, Warnings` (instead of: `Ok, Error`)
* I request no change in the `run-clang-tidy` exit code (but with 3 outcomes it may be a design option)

OTHERWISE: From my side
* The info on what `clang-tidy` detects or not, was informational
* You cannot do more than the `clang-tidy` tool does
* `run-clang-tidy` does what its name says, meaning: It runs `clang-tidy` (as `clang-tidy` wrapper)
* Adding additional checkers to your tool was not intended

PRAISE:
* What I really like about your `run-clang-tidy` tool is that it makes it easy to specify the file paths that are of interest in the config-file (similar to `clang-tidy` on the command line). That is a „unique selling point“ in my opinion compared to other `clang-tidy` runners.
* And the tool runs in parallel

HINT: 
The official clang-tools-extra project also provides a `run-clang-tidy` tool. That may cause a name clash with your tool and may cause problems if both are in the search path (`PATH` environment variable; a specific ordering is required). I assume you are aware of that.

**Comment by lmapii** [comment]
Ok, yeah it was also my intention to add the `Warning` to the existing `Ok` and `Error` indicators and then dump the warning at the end of the execution with a differently colored header than errors.

Sadly, I was not really aware of `run-clang-tidy` within the `clang-tools-extra` project. I started with https://github.com/lmapii/run-clang-format and simply added the same concept to this wrapper for `clang-tidy` ... May be worth considering renaming it to `rip-clang-format` and `rip-clang-tidy`, similar to https://github.com/BurntSushi/ripgrep (or just `r-clang-format`, etc.). Oh, well ;)

I'll look into it soon!

**Comment by lmapii** [comment]
I've pushed `0.2.0` which implements the `Warning` output and also adds the `--suppress-warnings` flag in case the user decides to ignore pure warnings. See [0.2.0 release](https://github.com/lmapii/run-clang-tidy/releases/tag/v0.2.0).

**Comment by jenisys** [comment]
Nice, I tested it with the new `0.2.0` version.
I looks good if I use `run-clang-tidy --supress-warnings ...` (which means: I see the progress bar w/ outcomes and the errors only).

**BUT: `run-clang-tidy -q ...` has a minor problem** (I think)
It shows you each error twice: 

* Once a block coloured in RED (with all files-with-errors), 
* then the results block where only the file-with-error ins marked with RED background color 
  (that you get with `--suppress-warnings` ).

**Comment by lmapii** [comment]
True. Sadly the crate I'm using doesn't have a "quiet" option but only log levels, so the fix is a bit hacky but nonetheless part of [0.2.1](https://github.com/lmapii/run-clang-tidy/releases/tag/v0.2.1).

**Comment by jenisys** [comment]
FIXED BY: v0.2.1
